Update config-settings.rst (experimental audit settings)

justinegeffen opened the pull request:
First draft of docs for https://github.com/mattermost/mattermost-server/pull/14062

wiggin77 commented:
Hi @justinegeffen ,  I think you asked me for a draft earlier and I didn't get to it.  Apologies for that.  He is my take, which you can change as needed:

The audit settings output audit records to syslog (local or remote server via TLS) and/or to a local file. Both are disabled by default.

Syslog configuration options:

SysLogEnabled: when “true” syslog output is enabled.
SysLogIP: IP address or domain of syslog server. Use “localhost” for local syslog.
SysLogPort: port the syslog server is listening on.
SysLogTag: syslog tag field; meaning is user defined.
SysLogCert: path to syslog server certificate for TLS connections (.crt or .pem).

File configuration options:

FileEnabled: when “true” file output is enabled.
FileName: path to output file.
FileMaxSizeMB: maximum size in megabytes file can grow before triggering rotation.
FileMaxAgeDays: maximum age in days before triggering rotation.
FileMaxBackups: maximum number of rotated files kept; oldest is deleted first.
FileCompress: when “true” rotated files are compressed using gzip.

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+
+File compress
+^^^^^^^^^^^^^^
+This when “true” rotated files are compressed using ``gzip``.
```
@wiggin77, what's the definition of "true"? Is it based on any of the other configurations or is it separate?

wiggin77 replied:
FileCompress is a boolean;  it can be `true` or `false`.   Like the other FileXXX config options, it is only used when `FileEnabled` is `true`.

justinegeffen replied:
That makes sense, thanks. So then ``FileCompress`` is not numerical, it would be ``true`` or ``false``?

wiggin77 replied:
Correct

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+
+File max backups
+^^^^^^^^^^^^^^
+This is the maximum number of rotated files kept; oldest is deleted first.
```
@wiggin77, is there a limit on the value that can be entered? Is there a default value?

wiggin77 replied:
Ah yes, I left out some info.  

FileMaxSizeMB defaults to 100
FileMaxAgeDays defaults to 0;  0 means no limit on age
FileMaxBackups defaults to 0;  0 means no limit on number of backups

wiggin77 replied:
None of those have any meaningful maximum value.

justinegeffen replied:
Thanks!

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+
+File max size MB
+^^^^^^^^^^^^^^
+This is the maximum size (measured in megabytes) that the file can grow before triggering rotation. The default setting is 100.
```
@wiggin77, is there a maximum value here?

wiggin77 replied:
No maximum other than what an Int can hold.  I don't think a maximum needs to be documented as it would exceed the size of any customer's disk capacity (something like 4 billion megabytes).

justinegeffen replied:
That makes sense, thank you. :)

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the path to the output file location.
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileName": []`` with string array input consisting of a user-defined name.                     |
```
```suggestion
| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileName": []`` with string array input consisting of a user-defined path.                     |
```

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the path to the output file location.
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileName": []`` with string array input consisting of a user-defined path.                     |
```
@wiggin77, would it be useful to have an example path here?

wiggin77 replied:
For Linux an example path would be  `/var/log/mattermost_audit.log`

wiggin77 replied:
Also, it is not an array of strings, just a single string.   Same for all the audit settings.

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+
+Syslog port
+^^^^^^^^^^^^^^
+The IP address or domain of the syslog server. Use ``localhost`` for local syslog.
```
```suggestion
The port that the syslog server is listening on.
```

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+The port that the syslog server is listening on.
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogIP": []`` with string array input consisting of an IP address or domain name.            |
```
```suggestion
| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogPort": []`` with string array input consisting of a port number.            |
```

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+The port that the syslog server is listening on.
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogPort": []`` with string array input consisting of a port number.            |
```
@wiggin77, is there a default port that should be used?

wiggin77 replied:
The default is 6514.   It is a numeric (not string).   Also, should the JSON 
```json
"ExperimentalAuditSettings": {
    "SysLogPort": 6514,
}
```
become `ExperimentalAuditSettingsSysLogPort` with no separator?  I've not worked with RST before so I don't know how that gets reformatted.

justinegeffen replied:
I'll review and revert on that.

justinegeffen replied:
AFAIK there is not meant to be a separator (based on existing content), but I am tagging @amyblais to confirm.

wiggin77 replied:
Reviewing the document, `ExperimentalAuditSettings` is the only section where the section name and key are joined like that.  As a user reading the doc, I would think `ExperimentalAuditSettingsSysLogPort` is a top-level key, but that wouldn't work.

justinegeffen replied:
Thanks, @wiggin77. I did a bit more digging into this and I based the format on the existing experimental config settings that are prefixed by "Experimental". I've updated the docs but was wondering if it's something that would be considered in a future release (appending "ExperimentalAuditSettings" to the setting)?

wiggin77 replied:
From what I see in config.json, when there is a single experimental option within an existing feature or section then the key is prefixed with `experimental`, such as `ExperimentalGroupUnreadChannels` within `ServiceSettings`.   When a larger feature is experimental it gets it's own section and the section is prefixed `experimental`.

justinegeffen replied:
Ah! Yeah that makes sense, thanks for looking into it. I've made the changes - thank you!

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+
+Syslog tag
+^^^^^^^^^^^^^^
+The IP address or domain of the syslog server. Use ``localhost`` for local syslog.
```
```suggestion
The syslog metadata tag field.
```

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+The syslog metadata tag field.
+
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogTag": []`` with string array input consisting of a user-defined field tag.             |
```
```suggestion
| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogTag": []`` with string array input consisting of a user-defined tag field.             |
```

justinegeffen commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the path to the syslog server certificate for TLS connections (``.crt`` or ``.pem``).
+
++---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogCert": []`` with string array input consisting of a user-defined field tag.             |
```
```suggestion
| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogCert": []`` with string array input consisting of a path to the location of the certificate.             |
```

justinegeffen commented:
Thanks, @wiggin77, that was exactly what I needed. I've updated the docs and added some comments around aspects that aren't immediately clear to me. Let me know if I have covered everything in terms of what's part of this release. Thanks. :)

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+**False**: Syslog output is disabled.
+
++-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogEnabled": "false"`` with options ``true`` and ``false``.                              |
```
I don't think the section and key should be combined here.  The section is `ExperimentalAuditSettings` and the key is `ExperimentalAuditSettingsSysLogEnabled`.  Also, the default value is a bool not a string.  e.g. ``"SysLogEnabled": false``

justinegeffen replied:
Updated, thanks!

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+The IP address or domain of the syslog server. Use ``localhost`` for local syslog.
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogIP": []`` with string input consisting of an IP address or domain name.                  |
```
Default value should be "localhost".  e.g. ``"SysLogIP": "localhost"``

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+The port that the syslog server is listening on. The default port is 6514. 
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogPort": []`` with numeric input consisting of a port number.                              |
```
Default here is 6514.  e.g. ``"SysLogPort": 6514``

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+The syslog metadata tag field.
+
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogTag": []`` with string input consisting of a user-defined tag field.                   |
```
Default is empty string, not empty array.  e.g. ``"SysLogTag": ""``

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the path to the syslog server certificate for TLS connections (``.crt`` or ``.pem``).
+
++---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsSysLogCert": []`` with string input consisting of the path to the certificate.                |
```
Default is empty string

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+**False**: File output is disabled.
+
++-------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileEnabled": "false""`` with options ``true`` and ``false``.                               |
```
Default is `false` not string `"false"`

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the path to the output file location.
+
++-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileName": []`` with string input consisting of a user-defined path (e.g. ``/var/log/mattermost_audit.log``).           |
```
Default is empty string.

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the maximum size (measured in megabytes) that the file can grow before triggering rotation. The default setting is 100.
+
++---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileMaxSizeMB": []`` with numerical input.                                                    |
```
Default is `100`.  e.g.  ``"FileMaxSizeMB": 100``

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the maximum age in days a file can reach before triggering rotation. The default value is 0, indicating no limit on the age.
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileMaxAgeDays": []`` with numerical input.                                                    |
```
Default is `0`

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This is the maximum number of rotated files kept; oldest is deleted first. The default value is 0, indicating no limit on the number of backups.
+
++----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileMaxBackups": []`` with numerical input.                                                    |
```
Default is `0`

wiggin77 commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4037,102 @@ Changes to this setting require a server restart before taking effect.
+This when “true” rotated files are compressed using ``gzip``.
+
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileCompress": []`` with options ``true`` and ``false``.                                     |
```
Default is `false`.  e.g.  ``"FileCompress": false``

amyblais commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4035,102 @@ Changes to this setting require a server restart before taking effect.
+Syslog configuration options
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+**True**: When ``true`` syslog output is enabled.
```
```suggestion
**True**: When ``true`` syslog output is enabled.

```

amyblais commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4035,102 @@ Changes to this setting require a server restart before taking effect.
+File configuration options
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+**True**: When ``true`` file output is enabled.
```
```suggestion
**True**: When ``true`` file output is enabled.

```

amyblais commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4035,102 @@ Changes to this setting require a server restart before taking effect.
+This is the maximum size (measured in megabytes) that the file can grow before triggering rotation. The default setting is 100.
+
++---------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| This feature's ``config.json`` setting is ``"ExperimentalAuditSettingsFileMaxSizeMB": 100`` with numerical input.                                                    |
```
This is not rendering.

justinegeffen replied:
Fixed.

wiersgallak commented on source/administration/config-settings.rst:
```diff
@@ -4037,6 +4035,104 @@ Changes to this setting require a server restart before taking effect.
+
+Syslog configuration options
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
Wondering if this needs a description?

justinegeffen replied:
I think both syslog and file config options would benefit from a description; I'll ping you when I've added them. :)

justinegeffen replied:
@wiersgallak, I've added descriptions.

justinegeffen commented:
@wiggin77, I've updated the docs as per our conversation yesterday around the additional settings. LMK if you're happy with the changes. Thanks!
